### DISCUSSION TITLE
Use a Project as an alternative to Milestones

### DISCUSSION THREAD

**Discussion by DNin01** [ScratchAddons/ScratchAddons#6618] [Discussion]
Calling all @ScratchAddons/admins

It seems like recently there has been a little more interest regarding the Projects feature within the ScratchAddons organization, based on what's been going on over at the Projects tab (I'm not going to get into detail or count them because some of them are private), so I was wondering if you all felt like creating a new Project to collect to-dos for releases instead of just using Milestones.

### Advantages

- Customizable layout
- Customizable fields (that we can create for things like status, due date, and version)
- Can include issues and PRs from other repositories, such as `website-v2`
- Talking planning and triaging, not much has to change and it doesn't have to be super complex, which should help us not burn out maintaining it

### Disadvantages

- Potentially slower access
- Some information may not be available
- Certain automations in the Project might be difficult

I like organizing, and one of my favorite parts is getting creative with it. I have tried out Projects before, so if anyone wants, I can make one _if you want_, and you can tell me what you think, but first, I think we should discuss it a bit so we can all agree on a layout and setup.

**Comment by Hans5958** [comment]
> so I was wondering if you all felt like moving away from Milestones into a new Project to collect to-dos for releases.

There are some that have been discussed on Discord (generally it is invoked because I wanted to create two projects for [website](https://github.com/orgs/ScratchAddons/projects/4) and [docs](https://github.com/orgs/ScratchAddons/projects/5) to let people know what is up), but the main point regarding this one is that "Milestones has its own purpose and should be kept."

Generally milestones is something that has an end, like a version has a time for a release. A project is a task board to track the progress of the overall project, not just a general milestone. Both of these have it's own purpose and can work in tandem. There is no need to one to be "being replaced" or anything.

**Reply by DNin01** [reply]
> Generally milestones is something that has an end, like a version has a time for a release. A project is a task board to track the progress of the overall project, not just a general milestone.

[Iteration fields](https://docs.github.com/en/issues/planning-and-tracking-with-projects/understanding-fields/about-iteration-fields) are kind of like milestones but they’re inside projects.

**Comment by WorldLanguages** [comment]
Personally, I was looking to create a project board for the ScratchAddons/ScratchAddons repo, but I wanted it to be fully automated. If you've ever heard or used TweetDeck (by Twitter) you know what I'll mean. I was looking for a way to have many cards/columns that each matches a different search filter.
It looks like there's no straight-forward way to achieve that, so I gave up lol. We're stuck with opening many browser tabs, each with a different search term. I'm a bit surprised this doesn't exist yet, as it's a very simple idea.

**Reply by DNin01** [reply]
What search filtering exactly were you aiming for?

We can always add views to save filters.

![Project view with saved filter](https://github.com/ScratchAddons/ScratchAddons/assets/106490990/e41d6b3c-e5c5-4898-b3a3-cf54b932b0b1)

**Reply by WorldLanguages** [reply]
> What search filtering exactly were you aiming for?

The search filters we use to search issues and PRs, and filter by labels, milestone, etc. The point being, nothing would be manually editable in the project board, it would simply be a view of many search terms in one page.

**Reply by WorldLanguages** [reply]
If you've got any ideas tell me, but having to manually edit the issue/PR milestone, labels, and also in the project board, is having to do everything twice for no reason.

**Reply by DNin01** [reply]
> > What search filtering exactly were you aiming for?
> 
> The search filters we use to search issues and PRs, and filter by labels, milestone, etc. The point being, nothing would be manually editable in the project board, it would simply be a view of many search terms in one page.

You mean the ones on the [org README](https://github.com/ScratchAddons/.github/tree/master/profile/README.md)?

**Reply by WorldLanguages** [reply]
Yeah

**Comment by DNin01** [comment]
This would be useful mainly just to benefit from the Projects feature set. I believe [Iterations](https://docs.github.com/en/issues/planning-and-tracking-with-projects/understanding-fields/about-iteration-fields) are the closest thing to Milestones in Projects. Instead of using Milestones, we could name each iteration after a version number and assign an end date that can be reflected through the project, which would enable us to use the Roadmap view. That whole documentation page is worth a read if you're not already familiar with iteration fields. They look really interesting and I see potential.

How does that sound for an upgrade? Or do Milestones work fine?

**Reply by Secret-chest** [reply]
Projects can show the milestone for each issue.

**Reply by DNin01** [reply]
> Projects can show the milestone for each issue.

I know, I just wanted us to be able to set a due version for issues outside the browser extension repository (such as blog posts for the website - each repository has a different set of milestones) and can the Roadmap view on Projects pull due dates from milestones or do you have to use an iteration instead? I don't know, I haven't tried it.

**Reply by WorldLanguages** [reply]
I think milestones work fine, and if we want to provide another view on pending issues/PRs, we should automate it so it syncs to the milestone assigned to each issue/PR.

**Reply by DNin01** [reply]
I believe you can specify the qualifier `-no:milestone` in the auto-add workflow to include issues and PRs in any milestone. I have not found a way to pull due dates from milestones yet - we will probably need to create our own automation for that. But there is always the option of switching to iterations.

But we always extend the deadlines on our milestones anyway, so I think the due dates don't really have that much meaning, so personally, I'm fine without due dates. Do you like the idea of a Project that automatically collects items from milestones on this repository? It might make it easier to jump to what we need to review and whatnot and it's still easy to group and slice by milestone. Any reasons you would like to see milestone tasks incorporated into a project? Personally, I just want to have customized layouts to help with finding what to work on.

I have some ideas for such a project and could start one as an experiment, see what I can do, and see if we like it, if you wouldn't mind. Would that be OK?

**Reply by WorldLanguages** [reply]
> Personally, I just want to have customized layouts to help with finding what to work on.

Yeah, that's the point.

**Reply by DNin01** [reply]
I want to help out by creating a project of my own, but do you have any ideas on what sorts of views and filters it should have? I been trying to decide if I want to start one yet (or maybe I should just try and see how it turns out?).

The reason I'm all for including issues and PRs from _milestones in particular_ is because we have to stay within the limit of 1,200 items per project (in case you forgot). I don't think it would be sustainable to include _all_ issues including those outside of milestones.

**Reply by WorldLanguages** [reply]
> I have some ideas for such a project and could start one as an experiment, see what I can do, and see if we like it, if you wouldn't mind. Would that be OK?

Sure, but if GitHub allows, try creating the board in your personal user instead of the org (I believe it was a mistake to do my tests in the org assuming it was not necessary).

> The reason I'm all for including issues and PRs from milestones in particular is because we have to stay within the limit of 1,200 items per project (in case you forgot).

lol had no idea there was a limit, but we don't want more than 100 anyway

**Reply by WorldLanguages** [reply]
> I want to help out by creating a project of my own, but do you have any ideas on what sorts of views and filters it should have?

I think being able to see these lists in one screen would be useful:
- Pinned issues/PRs - we would pin them more-or-less manually, not worth it to automate keeping the pinned issues in sync with the "pinned" card - and GitHub has no way to pin a PR.
- Issues with recent activity
- PRs with recent activity
- Two or three cards of PRs in closest milestone, filtered by label (possibly new addon vs scope:addon vs others)

But these would only work if we automate them, and they wouldn't really be *cards* in the typical sense, because they would be sorted as if they were searches. As I said above, I think we'd find it more helpful to have a TweetDeck-like interface instead of a board that resembles a physical board with post-its.

**Reply by DNin01** [reply]
> Sure, but if GitHub allows, try creating the board in your personal user instead of the org (I believe it was a mistake to do my tests in the org assuming it was not necessary).

If I do that, I wouldn't be able to include issues or milestones from this repository, I don't think. Having access to those makes it easier to set up the layout.

This project will not be designed as a pure test, more like a practical experiment - I have a good feeling that we might use the project a bit, and we could tweak it in multiple ways if necessary, so I kind of wanted to create it within the organization. But I understand that you might be a bit skeptical about the idea after having abandoned some tests (I might be a little bit, too), but I'll only create one.

My ideas are:
- Auto-add items from milestones on `ScratchAddons/ScratchAddons`
- Have a "recents" view sorting issues and PRs by recently updated (or recently created if you would prefer)
- A "Pinned" field - setting this would add an item to a "Pinned" view. In my personal to-do list project, I have an "Up Next" option in the Status field I use to highlight things I want to do next, so maybe that, to help with planning?
- Definitely grouping, slicing, or sorting by milestone (it doesn't seem like you can column by milestone on the board layout). Sorting should let us put the milestone for the nearest release at the top.
- An "attention needed" view which would show items with certain labels such as https://github.com/ScratchAddons/ScratchAddons/labels/status%3A%20awaiting%20answer%2Ffollowup, https://github.com/ScratchAddons/ScratchAddons/labels/type%3A%20security, etc.
- A way to see items assigned to you and potentially unreviewed/reviewed/yet to be triaged
- I don't want us to have to change the "same" thing in more than one place, so this one would focus on automation, with us intervening to pin and up-next items.

Any feedback? Let's get started?

**Reply by WorldLanguages** [reply]
> If I do that, I wouldn't be able to include issues or milestones from this repository, I don't think. Having access to those makes it easier to set up the layout.

Then feel free to create the project within the GitHub org. I guess you have enough perms to do so.

**Reply by WorldLanguages** [reply]
> A way to see items assigned to you

Is that even possible?

**Reply by WorldLanguages** [reply]
> Have a "recents" view sorting issues and PRs by recently updated (or recently created if you would prefer)

I think recently updated makes the most sense. Recently created are going to be recently updated anyway. Maybe from this view we can exclude things that are already at the top of other cards.

> Any feedback? Let's get started?

You are welcome to experiment with a private project (private as in, only visible to org members), I'm not sure exactly how these could be automated and what perms/keys you'll need.

**Reply by WorldLanguages** [reply]
Sorry for going off-topic, but I have a question: do yall reading this use github.com/notifications or do you actually receive GitHub notifications via email? (assuming you watch the SA/SA repo)

**Reply by DNin01** [reply]
> > If I do that, I wouldn't be able to include issues or milestones from this repository, I don't think. Having access to those makes it easier to set up the layout.
> 
> Then feel free to create the project within the GitHub org. I guess you have enough perms to do so.

Alright! Got somewhere to be, so I'll begin later today.

> > A way to see items assigned to you
> 
> Is that even possible?

`assignee:@me`

**Reply by DNin01** [reply]
> Sorry for going off-topic, but I have a question: do yall reading this use github.com/notifications or do you actually receive GitHub notifications via email? (assuming you watch the SA/SA repo)

I don't think I have it set to receive via email - I typically read notifications on the website and mobile app. Also, I only watch pull requests, releases, and security advisories at the moment.

**Reply by WorldLanguages** [reply]
I watch everything 💀

**Reply by WorldLanguages** [reply]
> > > A way to see items assigned to you
> > 
> > 
> > Is that even possible?
> 
> `assignee:@me`

How would that work in a project board that can be viewed by more than 1 user? 🤔 
AFAIK you can't directly include a search term widget in a GitHub project board. Or will this not be a GitHub project board?

**Reply by BroJac5246** [reply]
> Sorry for going off-topic, but I have a question: do yall reading this use github.com/notifications or do you actually receive GitHub notifications via email? (assuming you watch the SA/SA repo)

I watch everything and turned off email notifications. I mostly use the app on my phone, and I use the website on my laptops.

**Reply by DNin01** [reply]
> > `assignee:@me`
> 
> How would that work in a project board that can be viewed by more than 1 user? 🤔 AFAIK you can't directly include a search term widget in a GitHub project board. Or will this not be a GitHub project board?

[`@me` automatically selects the currently logged-in user](https://docs.github.com/en/issues/planning-and-tracking-with-projects/customizing-views-in-your-project/filtering-projects#filtering-assignees-and-reviewers-using-keywords). If I wanted to make a view showing issues assigned to you, I can just type that into the filter box and save the view.

**Reply by DNin01** [reply]
Looks like [there's an incident with the Issues service](https://www.githubstatus.com/) affecting Projects. After that's resolved I'll try making a project.

**Reply by DNin01** [reply]
As it turns out, it is not possible to specify `-no:milestone` in an auto-add workflow. That means we have to logically OR with the milestones we've currently created with `milestone:v1.35.0,v1.36.0`etc.. That'll require a bit of maintenance, but I guess we can still make it work. Alternatively, we could start using a "todo" label on issues/PRs. IDK.

**Reply by DNin01** [reply]
Whew, making a project yours is harder than you think, and I'll have to spend some more time setting up the layout for an ideal experience. The project is ready and available to organization members on the [organization Projects section](https://github.com/orgs/ScratchAddons/projects).

I'm currently wondering if this is even the kind of thing we were looking for, but maybe time will tell.

@WorldLanguages Any thoughts now? Over the next few days, I'll try to spend some time with contributor experience in mind as I tweak the project.

**Reply by DNin01** [reply]
I haven't added a "pinboard" yet and I'm still considering if we need it. At the moment, I have an "Up Next" option for Status which kind of functions like a pinboard for issues and a "review this next" section for PRs. I don't want a pinboard to leave us questioning what should be pinned or divert our attention away from less important but still important items, which is mainly why I haven't made it yet, aside from "Up Next".

**Reply by Secret-chest** [reply]
Email only.

**Reply by WorldLanguages** [reply]
> Any thoughts now?

Looks good I think, maybe the "explore" view should be renamed? I didn't know what it was showing until I looked at the search filter.

**Reply by DNin01** [reply]
It's fun watching the items get moved between groups and columns in real time 👀

**Comment by DNin01** [comment]
Would it be possible to link the project I made to this repo for quick access? The Projects tab will need to be enabled on this repository.

Just means it'll take less clicks to get there, but it's still in its "experimental" phase, so I don't really care if it's linked or not.

**Reply by DNin01** [reply]
Potentially some other projects too, if we feel like they fit.

**Reply by WorldLanguages** [reply]
We could also link to them from the org readme

**Reply by Hans5958** [reply]
I have activated the Projects feature. Sorry for the wait.

**Comment by DNin01** [comment]
Well, now that it's had some time, what do you think of the To-Do Table project I made a while ago, @ScratchAddons/admins? I imagine it replacing our Backlog milestone and maybe even priority labels (if applicable; I'm not exactly sure how our priority labels were meant to be used). Since it's a project, we can also keep the board private for just us to see and use, but I think it would be nice to make it public, especially if it replaces anything.

What do you think of our Backlog milestone? Milestones typically have an end date, but this one doesn't really, making the completion percentage useless. Other alternatives would be just using priority labels or making a "to do" label. But Projects can make it easier to see what tasks we need to do and when, if we use it as a central location to display them. Plus, we can sort and slice them, save views... various nice things for our workflow. It's a better fit for a backlog and could encourage us to add items as to-dos more frequently, so I propose dropping the Backlog milestone and switching to using this project.

I don't know if it solves @WorldLanguages's [suggestion](https://github.com/ScratchAddons/ScratchAddons/discussions/6618#discussioncomment-6897541), but it still seems like an improvement to our workflow.

Of course, we could also use both! Milestones _are_ part of the grouping in the project I created. At this point, I think we should keep the version-number milestones we're currently using and integrate them into the project, unless there's a good reason to switch to iteration fields within the project.

**Reply by DNin01** [reply]
Just so you know...
![Visibility setting set to Private with description reading, "Only organization owners can change project visibility"](https://github.com/ScratchAddons/ScratchAddons/assets/106490990/de9ab1bf-396a-4760-ab47-1cac3d0faa5a)

Would you, the admins, be okay with it being public? Most of the items are from milestones. I'd be fine with it being public once it replaces our Backlog milestone (the milestone doesn't have to be closed or deleted right away; I just mean you can make it public when we commit to transitioning from the Backlog milestone to the project).

**Reply by WorldLanguages** [reply]
I think it would be nice to replace the backlog milestone with a public project

**Reply by WorldLanguages** [reply]
Can you send a link to your project? We have too many projects named "test" etc. 😅 we would also want to get rid of most of those

**Reply by DNin01** [reply]
> Can you send a link to your project?

Yes, here it is: https://github.com/orgs/ScratchAddons/projects/11

**Reply by DNin01** [reply]
Here's everything from the Backlog milestone that is not included in the project:

https://github.com/ScratchAddons/ScratchAddons/issues?q=is%3Aopen+milestone%3ABacklog+-project%3Ascratchaddons%2F11

One of the next steps is to re-assess everything in the Backlog milestone and add whatever we're still planning to the project. We can then delete or close the Backlog milestone.

**Reply by Hans5958** [reply]
The Backlog milestone is manual, so I would expect that the new Backlog project is also manual, whereas the current To-Do List test project is automated.

**Reply by DNin01** [reply]
> The Backlog milestone is manual, so I would expect that the new Backlog project is also manual, whereas the current To-Do List test project is automated.

Actually, at this time, it's not fully automated. As I kept experimenting, I've kind of turned the to-do list test project from a milestone sorter into a general, manual to-do list (with automated milestone integration on top), so yeah, it's essentially our new Backlog project.

**Reply by DNin01** [reply]
I mean, the project doesn't _really have_ to be public. We could instead use our milestones to represent our public roadmap and keep the to-do table private, unless the team is okay with the public being able to see our plans in detail.

It's kind of up to you. I know you're the admins, so I try to only add tasks if I'm at least 90% sure they're planned. I'm not sure how many of you have interacted with the to-do table.

**Reply by WorldLanguages** [reply]
I think you may be overestimating the "admin" role, it just means admin permissions on GitHub.
It's hard to define what "planned" means on this project, unless it's something like upgrading to MV3. As long as it's not understood as a guarantee of something ever getting done, you'll be okay at judging what's planned/accepted.

**Reply by WorldLanguages** [reply]
Who knows where this project would be right now if I was more focused on planning, but sometimes you can't have it all. I guess I was never fully ready to get so many contributions and see our user count increase 100x.
